src/Cake/Scripting/Roslyn/RoslynScriptSession.cs

```diff
@@ -78,62 +91,132 @@ public void ImportNamespace(string @namespace)
+                                break;
+                            case DiagnosticSeverity.Error:
+                                _log.Error(diagnostic.ToString());
+                                errors.Add(diagnostic);
+                                break;
+                            default:
+                                break;
+                        }
+                    }
+
+                    if (errors.Any())
+                    {
+                        var errorMessages = string.Join(Environment.NewLine, errors.Select(x => x.ToString()));
+                        var message = string.Format(CultureInfo.InvariantCulture, "Error(s) occurred when compiling build script:{0}{1}", Environment.NewLine, errorMessages);
+                        throw new CakeException(message);
+                    }
+                }
+                if (_scriptCacheEnabled)
+                {
+                    // Verify cache directory exists
+                    if (!IO.Directory.Exists(_scriptCachePath.FullPath))
+                    {
+                        IO.Directory.CreateDirectory(_scriptCachePath.FullPath);
+                    }
+                    if (string.IsNullOrEmpty(scriptHash))
+                    {
+                        scriptHash = FastHash.GenerateHash(Encoding.UTF8.GetBytes(string.Concat(script.Lines)));
+                    }
+                    var emitResult = compilation.Emit(cachedAssembly.FullPath);
+
+                    if (emitResult.Success)
+                    {
+                        IO.File.WriteAllText(hashFile.FullPath, scriptHash);
+                        RunScriptAssembly(cachedAssembly.FullPath);
+                    }
+                }
+                else
+                {
+                    using (new ScriptAssemblyResolver(_log))
+                    {
+                        roslynScript.RunAsync(_host).Wait();
+                    }
                 }
             }
+        }
 
-            if (errors.Any())
-            {
-                var errorMessages = string.Join(Environment.NewLine, errors.Select(x => x.ToString()));
-                var message = string.Format(CultureInfo.InvariantCulture, "Error(s) occurred when compiling build script:{0}{1}", Environment.NewLine, errorMessages);
-                throw new CakeException(message);
-            }
-
+        private void RunScriptAssembly(string assemblyPath)
+        {
+            var assembly = Assembly.LoadFile(assemblyPath);
+            var type = assembly.GetType("Submission#0");
+            var factoryMethod = type.GetMethod("<Factory>", new[] { typeof(object[]) });
             using (new ScriptAssemblyResolver(_log))
             {
-                roslynScript.RunAsync(_host).Wait();
+                try
+                {
+                    var task = (System.Threading.Tasks.Task<object>)factoryMethod.Invoke(null, new object[] { new object[] { _host, null } });
```

Comment:
No need for `System.Threading.Tasks`. Include the namespace instead.

Reply:
Fixed